### DISCUSSION TITLE
Logging the Solving Process for Step-Wise Analysis (Employee Scheduling Problem)

### DISCUSSION THREAD

**Discussion by G-Phong** [google/or-tools#3899] [CP-SAT questions]
Hello OR-Tools community and @lperron,

I am working on a shift scheduling problem very similar to the Employee scheduling example provided [here](https://developers.google.com/optimization/scheduling/employee_scheduling). In my application, I'm implementing a preference system for shifts based on numerical values ranging from 0 (no preference) to 100 (full preference), replacing the shift request mechanism.

My aim is to capture the solving process in a step-wise manner to understand the chronological order of assignments, specifically which variables are assigned first and which ones later. The goal is to log the decisions made during the solving process, enabling a transparent visualization for human-centered explanations. Is it even **possible** to explain the Solver-decisions in a step-wise manner?

I am interested in exploring the generation of natural language explanations that describe the reasoning behind the decisions made, therefore I think I will need to have these log information. These explanations would take the form of sentences like, “From the preferences X and constraints Y, and considering the fact that certain conditions are met (e.g., shift requirements, employee availability, shift assignments which are already made), it follows that these specific shift assignments are made.”

I have read the discussion [here](https://github.com/google/or-tools/discussions/2427) and understand that the .solve() function itself is _stateless_. I'm looking for suggestions on how I might still observe and log the assignment process.

Any insights or references to relevant documentation would be greatly appreciated. Your expertise in this area could be instrumental in making the algorithm more explainable and transparent to the users/workers.

Thank you for your time and assistance.

Best regards,
Phong (writing his master thesis about this)

**Comment by sschnug** [comment]
An opinion:

I would be worried, that this kind of information is way too verbose, uncompressed and chaotic.
(I would not be surprised so see a 10GB output for a 5 sec solver run tracking all the stuff happening)

Even if you would start with a classic CP solver, the amount of decisions and backtrackings made will easily overpower the human.

If you are using CP-SAT howewer, it's way worse in that regard: 

- a more decomposed boolean core-model
  - where variables are even created on the fly
- non-chronological backtracking
- an immense amount of heuristics
- some hard to grasp stuff like reduced-cost fixing, cutting-planes and other LP-relaxation based things with a global-view (hard to follow)

It's extremely complex.

Maybe it's less about the solving process, but about evaluating solutions compared to neighbors in a sensitivity-analysis like manner?
-> Your mouse hovers over assignment x/y -> the next k-best solutions where this assignment is NOT done look like and those solutions differ in their trade-offs ...

To phrase it differently:

Imho it's not about how the solver did find a solution, but what the solver can prove about this solution (compared to some alternatives) in regards to what some potential user is interested in in some interactive setting. 
With this approach, much more human-compatible things are possible like compressing / compacting proofs (think minimal unsatisfiable subsets and co.).

(It's highly linked to incremental-solving and infeasibility/lower-bound proofs).

I'm pretty sure that there is some research into this but i'm unable to list keywords. Maybe your academic advisor has some insights?

**Reply by G-Phong** [reply]
First of all, thank you @sschnug so much for your quick and comprehensive response!

I agree with your concerns about tracking the entire solving process, especially with the CP-SAT solver. I've looked through the source code of the solve() function and I understand that it may not necessarily translate into user-friendly or meaningful explanations. Also I doubt that I am capable of implementing this functionality by myself as the CP-SAT-Solver algorithm is pretty complex...

Given your thoughts, I am inclined to pursue **a more contrastive explanation approach**, i.e. comparing alternatives, implementing a "What-If-Tool", or doing some kind of sensitivity analysis by e.g. changing employee availabilities, changing skill requirements or also changing weights of individual employee preferences. I believe this direction could lead to more human-centered explanations by showcasing alternative assignments and contrasting them with the chosen solution by comparing preferences. This way, users may find it easier to grasp the reasoning behind the algorithm's decisions.

I still got two questions concerning your answer.

> With this approach, much more human-compatible things are possible like compressing / compacting proofs (think minimal unsatisfiable subsets and co.).

1) What do you mean with compressing/compacting proofs? I am not an expert when it comes to concepts like minimal unsatisfiable subsets... also in my application I assume that 99% of employee scheduling problems are at least feasible...


> (It's highly linked to incremental-solving and infeasibility/lower-bound proofs)

2) What do you mean with this? Is this a technique to explain infeasible problems?

**Reply by CADBIMDeveloper** [reply]
@G-Phong ,

An opinion:

In the original employee scheduling example no model objective is provided and the solver just enumerate possible (feasible) solutions. What if you add an objective ("employee total satisfaction" :-) ) to this problem. You can also track intermediate solutions using a solver solution callback. In that case you can collect multiple solutions and now you have a score of each one.

Once you've got an optimal solution you can make this objective value to be a constraint for a new search (and probably constrain employees schedules without preferences) and run this new task without an objective to get alternative schedule.

What do you think?

**Reply by G-Phong** [reply]
@CADBIMDeveloper, thank you for sharing your thoughts and nice ideas

Indeed, I've already incorporated an objective into the problem, which is the total preference of all employees (employee preferences are given as numerical values from 0-100 and for each allocated shift the preference value is summed up). This represents my version of "employee total satisfaction" you mentioned.

I've also already implemented the solution callback mechanism you suggested, and it's working well. When multiple optimal solutions exist, I can now showcase all of them and put them in contrast.

Your idea about using the objective value as a constraint for a new search to find good but not necessarily optimal solutions is very interesting. This could indeed help if I only obtain 1 optimal solution, then the solution callback mechanism does not show other alternative solutions... then I could "manually" generate good solutions using your approach, right? I should probably also not put the exact objective value into the new constraint but something like 80% or 90% of that objective value, so I can generate "good" but not optimal solutions.

I will try that out and see how it performs in practice.

**Reply by CADBIMDeveloper** [reply]
@G-Phong 

Yes if the solver has an objective, then it doesn't iterate and report other possible solutions.

I would also recomment to constrain employee shifts without preferences to reduce alternative solutions count

**Reply by gregy4** [reply]
Few years ago existed application cpviz for visualization of cp decision trees. It worked offline, at beginning you stored decision tree and variable domains at every world index in xml files and then visualized them in cpviz.
For my problem, visualization didn't worked (too large problem) but as the last possibility it worked for me to look at a decision tree. I have to agree this is very hardcore approach and I was on a limit of my brain. On other side I have unique experience, I know deeply how cp solver works.
Cp-sat solver uses different approach that is not trackable as cp solver. Maybe combination of both solvers where cp-sat can prove infeasibility of some paths can work or looking at the results and thinking about them as other suggested.

**Reply by G-Phong** [reply]
@gregy4, I have read a little bit about that cpviz application. I doubt that it contributes to "human-centered explanations" as of its complex output. This might be more suitable for technical experts and for those who want to understand the solving process (like you). I also understand that CP-SAT is way different from a "usual" CP-Solver...

**Reply by gregy4** [reply]
Agree, cpviz output should be somehow compressed. For example repeating failures with change variable domains by 1. I though about fixed part of decision tree (previous decisions that weren't touched by backtracking) and currently closed branch. You can say that for decisions in fixed part was found a better solution in the branch or not and characteristics of new branch you want to investigate. Of course it is important whether variables in your models make sense to users or not and whether used branching strategy is fast and easy  to understand (not start with last decision with a lot backtracking).

**Comment by lperron** [comment]
Totally agree
